feat/updateArticle : 아티클 업데이트 하기 (UPDATE), 관련 리팩토링

katej927 opened the pull request:
# What is this PR?

1. 아티클 업데이트 하기 (UPDATE)
2. 관련 리팩토링

# Changes

1. 아티클 업데이트 하기 (UPDATE)
    - 에러 처리
        - response의 data가 없는 경우 not-found 출현 시킴
    - use hook 사용
        
        clinet 컴포넌트에서도 async/await 사용 가능하도록 제작
        
2. 관련 리팩토링
    - api와 Interface를 한 곳(api 파일)에 몰아두기
    - 글 생성/수정 할 때 form의 UI 컴포넌트 (`articleForm`) 분리 및 재사용
    - alias 수정
        
        src/*, src/containers/* 를 모두 @로 import하여 가독성 향상
        

# Screenshot

프리뷰의 CORS 에러를 아직 해결하지 못했습니다. 시간이 너무 오래 걸려서 일단 스크린샷으로 설명 할게요.

|action |image|
|--|--|
|상세 페이지 진입 후, 수정 텍스트 클릭|<img src='https://github.com/f-lab-edu/dev-blog/assets/69146527/a722eae5-4fb1-433e-a899-eb4834b6e422' height='200'/>|
|edit 페이지 진입 해서 내용 수정 후, publish|<img src='https://github.com/f-lab-edu/dev-blog/assets/69146527/384cddc2-5454-4b99-bf56-a91226e82d57' height='200'/>|
|변경 후, 변경된 내용의 상세 페이지 라우팅됨.|<img src='https://github.com/f-lab-edu/dev-blog/assets/69146527/f2f6999e-3951-4a87-b3a8-f8a91c050530' height='200'/>|

# Questions

- EditArticle 컴포넌트
    - NotFound를 저렇게 넣어도 되는지?
    - use를 사용해도 되는지?
        
        사용 이유: handleSubmit을 내려줘야 하고 router를 사용해야 하는 client 컴포넌트에서 데이터를 가져오는 서버 기능을 넣어줘야 했음. async/await가 getArticleByIdAPI 내부에 있어서 에러가 나던 것을 use를 통해 해결했으나, 그 아래 코드에서 <NotFound />가 아닌 notFound()를 기입하면 에러 발생.
        
        해결을 했으나 잘 해결한 것인지에 대한 확신은 없어 질문드립니다. (시간이 이미 많이 소요되어서)
        
- 앞선 PR([feat/deleteArticle : 아티클 지우기 (DELETE), 관련 리팩토링](https://github.com/f-lab-edu/dev-blog/pull/5#top))에서 언급했듯이 ‘홈으로 이동 후 api를 refetch’구현을 아직하지 못했습니다.

# Trouble Shooting

※ PR별 유의미한 트러블 슈팅을 기록하여 개인적으로 참고 하고자 적습니다. 읽지 않으셔도 괜찮습니다.


<details>
        <summary>1. server/client component와 use hook</summary>
- src/app/editArticle/[id]/page.tsx

server와 client component를 처음 사용하게 되면서 사용할 수 있는 기능에 제약이 있는 것을 잘 알지 못했다.
백엔드에서 데이터를 가져오는 로직을 위해 async/await도 써야 하고 (server component 기능) router를 사용하고 함수도 내려줘야 하는데 (client component 기능) 모두 사용하려니 계속 에러가 나서 헤맸고 매우 답답했다.

전 회사에서라면 10분내로 답 안 나오면 바로 질문하라고 해서 사수님의 도움을 받으며 해결했을테지만 멘토링에서는 멘토님이 끝까지 찾길 추천하셨고 아니면 따끔하게 지적받기에 계속 찾아봤다.
꽤 오랜 시간이 걸렸고 포기하고 멘토님께 그냥 질문드려 해결하고 싶은 순간들도 있었으나, 돌아올 답이 무엇인지 알 것도 같고 그냥 한 번 찾아보았다. 개발자가 스스로 찾는 것도 중요하기 때문이다. (회사에서는 시간 관계 상의 이유로 빠르게 질문하는 것을 권유한 것으로 안다.)
아마도 에러메세지를 쳤고 원어를 해석해내는 고통을 견디면서 stackoverflow에서 우연히 use를 사용하면 해결 가능하다는 것을 봤고 그것이 내가 알지 못한 hook이며 근래에 추가 되었다는 것을 알게 되었다.
그래서 use hook에 대해 공식문서와 블로그에서 어떤 기능을 하는 hook인지 찾아보았고 내가 원하던 기능을 하는 알맞은 function임을 알았다.

뿌듯했고 성취감을 느꼈다. 얼마 전, preview 구현하면서도 이번에도 (비록 오랜 시간이 걸렸지만) 스스로 해결해냈다.

**단순히 끝날 줄 알았던 update와 delete였는데 새로이 업데이트된 기능으로 구현하는데 차질이 많았고 스스로 해결하려다 보니 오랜 시간이 걸렸으나 뿌듯했다.**
**얻은 점은 조금씩, 스스로 문제 해결하는 방법을 찾는 것 같다는 점이다. 에러 로그를 이해하고 부족하면 에러 로그를 검색하고 (원어라 고통스럽지만) issue탭이나 stackoverflow에서 나와 비슷한 문제를 겪는 이들이 있는 것을 확인하고 그들이 어떻게 해결해내고 있는지 알 수 있다.**
**앞으로도 이런 방식을 조금 더 빨리 실행해서 스스로 해결할 때 속도를 높이는 것이 좋을 것 같다.**
</details>

katej927 committed:
fix : replace localhost url to API_URL

katej927 committed:
feat : add editArticle router button & page

katej927 committed:
refactor : add articleForm for reusing in CREATE & UPDATE
- applied in addArticle page (CREATE)
- add ArticleInterface for reusing and applied

katej927 committed:
feat & refactor : revised name 'topic' -> 'article', making EditArticle page
- changed SSR to CSR (when it occured error)

katej927 committed:
feat : add 'use' hook to use CSR component
- instead removed async/await
- can deliver function to children component

katej927 committed:
feat & refactor : add NotFound page, revised duplicated logic(getArticleById)'s location into services folder, routing to article page when successfully edit

katej927 committed:
refactor : moved getArticles method & getArticle interface into service folder
- for readability

katej927 committed:
feat : add validation for edit Article
- not to submit with empty string

katej927 committed:
refactor : revised alias root and applied

katej927 committed:
refactor : revised name of api

katej927 committed:
fix : applied function that name changed

katej927 committed:
fix : fix wrong method name, replace <NotFound /> to notFound()

katej927 committed:
fix : add configs to fix CORS error in preview URL

katej927 committed:
fix : removed nextConfig, eslint setting

f-lab-zero commented:
> EditArticle 컴포넌트

> NotFound를 저렇게 넣어도 되는지?

> use를 사용해도 되는지?

> 사용 이유: handleSubmit을 내려줘야 하고 router를 사용해야 하는 client 컴포넌트에서 데이터를 가져오는 서버 기능을 넣어줘야 했음. async/await가 getArticleByIdAPI 내부에 있어서 에러가 나던 것을 use를 통해 해결했으나, 그 아래 코드에서 가 아닌 notFound()를 기입하면 에러 발생.

코드에 코멘트를 안달면 제가 왔다갔다하면서 봐야해서 리뷰하기 힘든 것 같아요 ㅠ

f-lab-zero commented on src/services/articles.ts:
```diff
@@ -0,0 +1,53 @@
+import { API_URL } from '../constants/common'
```
api들의 폴더명을 services로 한 이유가 무엇인가요? services 라는 폴더명을 보고 모르는 사람이 여기에 api 함수들이 모여있구나, 라고 파악하여 찾기는 힘든 것 같아요. 모두가 알기 쉬운 폴더명, 함수명, 변수명을 짓는게 가독성 향상의 첫 걸음같아요.

katej927 replied:
services로 한 이유는 레퍼런스를 모방한 건데요, 적어주신 내용을 보니 수정해야 할 것 같아서 변경해둘게요. 
(apis로 수정)

f-lab-zero commented on src/interface/article.ts:
```diff
@@ -0,0 +1,4 @@
+export interface ArticleInterface {
```
type을 구현체와 폴더까지 분리시킨 이유는 무엇인가요?
연관된건 가까이 두어야 보기 쉽고, 찾아가기 쉬워요.
간단해 보이지만 이런 것 만으로도 생산성을 높입니다.

f-lab-zero replied:
비슷한것, 연관된것들은 최대한 가까이 위치시켜주세요.

katej927 replied:
ArticleInterface는 api외에 다른 곳(ex. UI 쪽 props)에서도 많이 사용되기 때문에 구현체와 폴더까지 분리시켰는데요, 그냥 api interface 있는 곳에 합치는 게 좋겠다는 말씀이시겠죠?

적어두신 기준으로 보면 연관된 것들이 api interface에 많이 있기 때문에 api interface쪽에 넣는 게 맞아보이기도 해요.

f-lab-zero commented on src/services/articles.ts:
```diff
@@ -0,0 +1,53 @@
+  _id: string
+  createdAt: string
+  updatedAt: string
+  __v: number
```
`__v` 는 어떤거에요? 
클라에서 사용하는게 아니라면 굳이 서버에서 안넘겨줘도 좋을 것 같아요

katej927 replied:
어떤 건지 모르고 사용되지 않기에 삭제했습니다!

katej927 commented:
> > EditArticle 컴포넌트
> 
> > NotFound를 저렇게 넣어도 되는지?
> 
> > use를 사용해도 되는지?
> 
> > 사용 이유: handleSubmit을 내려줘야 하고 router를 사용해야 하는 client 컴포넌트에서 데이터를 가져오는 서버 기능을 넣어줘야 했음. async/await가 getArticleByIdAPI 내부에 있어서 에러가 나던 것을 use를 통해 해결했으나, 그 아래 코드에서 가 아닌 notFound()를 기입하면 에러 발생.
> 
> 코드에 코멘트를 안달면 제가 왔다갔다하면서 봐야해서 리뷰하기 힘든 것 같아요 ㅠ

코멘트 달고 슬랙에 쓰레드 다시 남겨둘게요! 가독성 생각을 못했어요 ㅠ

f-lab-zero commented on src/components/articleForm/index.tsx:
```diff
@@ -0,0 +1,47 @@
+
+import { ArticleInterface } from '@/interface/article'
+
+interface IProps extends ArticleInterface {
```
Props 앞에 I 를 붙힌건 인터페이스라는 뜻이죠? 없어도 될 것 같아요. 있어서 장점이 없는 것 같아요

ArticleInterface 도 똑같은 인터페이스인데 앞에 I가 안붙었는데 이유가 무엇인가요??

katej927 replied:
> ArticleInterface 도 똑같은 인터페이스인데 앞에 I가 안붙었는데 이유가 무엇인가요??

ArticleInterface에는 interface가 명시가 되어 있고 IProps는 명시되어 있지 않기 때문이에요. I는 type과 interface를 구분짓기 위함인데요, 언급해주셔서 생각해보니 굳이 앞에 T나 I를 붙여 구분지을 필요가 있을까? 싶은 생각이 들었어요. 구분 지을 상황이 없을 것 같아서요.

그래서 IProps를 Props로 사용하도록 할게요.

f-lab-zero commented on src/components/articleForm/index.tsx:
```diff
@@ -0,0 +1,47 @@
+  }: ChangeEvent<HTMLInputElement>) => setNewContent(value)
+
+  const handleSubmit = (e: FormEvent<HTMLFormElement>) =>
+    onSubmit(e, { title: newTitle, content: newContent })
```
event는 안넘겨도 될 것 같은데 넘기는 이유가 있을까요?

katej927 replied:
`e.preventDefault()` 를 위해서 넘겼습니다! 이를 위해서라면 필요하지 않을까요?

f-lab-zero replied:
e를 꼭 밖에서 처리해야 하나요? 여기서 처리하고 props로 넘겨받는 submit에는 안넘겨줘도 될 것 같아서요

katej927 replied:
오 맞네요! 여기서 처리하고 넘겨받는 곳에서는 지워줬어요.

f-lab-zero commented on src/services/articles.ts:
```diff
@@ -0,0 +1,53 @@
+export const getArticlesAPI = async (): Promise<
+  GetArticlesResponseInterface | undefined
+> => {
+  try {
```
api들의 try/catch를 api 함수내에서 모두 진행하고 있는데 만약 해당 함수를 가져다 사용하는 곳에서 api 에러 발생하면 에러 모달을 띄워주거나 다른 로직을 수행하려면 어떻게 해야할까요?? 에러 핸들링은 가져다 사용하는 곳에서 하는게 좋을 것 같아요

katej927 replied:
🙋‍♀️ 질문 : 

```ts
export const getArticleById = async (
  id: string,
): Promise<GetArticleResponseInterface | undefined> => {
  try {
    const res = await fetch(`${API_URL}/api/articles/${id}`, {
      cache: 'no-store',
    })

    if (!res.ok) {
      // throw new Error('Failed to fetch an article.')
      return 'error' // 👈
    }

    return res.json()
  } catch (error) {
    console.log(error)
  }
}
```

멘토님 위 코드처럼 getArticleById 함수의 return 에 에러를 구분할 수 있는 값을 넣어주고 호출한 곳에서 에러 인식이 되면 그 때 처리하면 되는 걸까요?

f-lab-zero replied:
이런 api 함수 내부에 try/catch가 있으면 안된다는걸 이야기 하고 싶었어요.
네트워크 에러등이 발생하면 catch에서 잡힐껀데 그러면 어쩌실건가요? 
추가로 여러 종류의 에러들은 404에러 500 에러등은 어떻게 구분시킬껀가요?

katej927 replied:
고민해보고 리팩토링할 때 반영할게요.

katej927 replied:
반영 완료

f-lab-zero commented on src/services/articles.ts:
```diff
@@ -0,0 +1,53 @@
+  }
+}
+
+export const getArticlesAPI = async (): Promise<
```
get 이라는것에서 api 의미를 내포하고 있어 뒤에 API postfix는 불필요해 보여요.

katej927 replied:
반영했습니다.

katej927 committed:
refactor : revised name of service folder
- 'service' -> 'apis'
- reason : for understandable to anyone
- applied to every files

katej927 committed:
refactor : moved ArticleInterface's location and applied

katej927 committed:
refactor : removed unused props('__v') from GetArticleInterface

katej927 committed:
Merge branch 'main' into feat/updateArticle

katej927 committed:
fix : solved conflicts

katej927 committed:
refactor : remove I from IProps, export and applied GetArticleInterface

katej927 committed:
refactor : remove 'API' postfix from get api methods
- reason : 'get' has meaning of 'api'

katej927 committed:
fix : import path error

katej927 committed:
fix : add edit button which is removed mistakenly

katej927 commented on src/app/[id]/page.tsx:
```diff
@@ -1,29 +1,13 @@
+  const data = await getArticleByIdAPI(id)
 
-  if (!data) return
+  if (!data) return notFound()
```
[PR 첫 번째 질문] NotFound를 이렇게 넣는 것이 맞는지? 1번안

f-lab-zero replied:
되지 않을까요?

f-lab-zero replied:
왜 의문이 들었나요?

katej927 replied:
'2번안'에서는 안되니까 그런 것 같습니다..! 안 써봐서 낯설기도 했는데 지금 생각해보니 '1번안'에서는 적절히 사용했다는 생각이 들어요. 유지할게요!

katej927 commented on src/app/editArticle/[id]/page.tsx:
```diff
@@ -0,0 +1,61 @@
+  const router = useRouter()
+  const data = use(getArticleByIdAPI(id))
+
+  if (!data) return <NotFound />
```
- PR 중 해당 되는 부분
  - Changes 
    - response의 data가 없는 경우 not-found 출현 시킴
  - Questions
    - 첫 번째 질문 : NotFound를 이렇게 넣는 것이 맞는지? 2번안
    
       여기서는 notFound()를 쓸 경우, 계속 에러가 나더라구요. 에러 내용은 14번째 줄에서 async/await(getArticleByIdAPI 함수 내부에서 사용)를 사용하면 안된다는 이유로요.

f-lab-zero replied:
에러가 왜나는지는 use를 잘몰라서(한 번도 써본적 없어요) 제대로된 답변을 못할 것 같네요. 선미님이 트러블 슈팅 해보시는게 좋을 것 같아요.

katej927 replied:
🙋‍♀️ 질문 : 
ㅠㅠ 그럼 이 부분은 다른 PR (리팩토링 등)에서 처리해도 될까요? 
여기서는 시간이 너무 많이 지체되었고 당장 해결하기 어려워 보여서요.

f-lab-zero replied:
넵

katej927 commented on src/app/editArticle/[id]/page.tsx:
```diff
@@ -0,0 +1,61 @@
+
+const EditArticle = ({ params: { id } }: { params: { id: string } }) => {
+  const router = useRouter()
+  const data = use(getArticleByIdAPI(id))
```
- PR 중 해당 되는 부분
  - Changes 
    - use hook 사용
  - Questions
    - 두 번째 질문 : use를 사용해도 되는지?
       설명은 두번째 질문에 있는 '사용 이유'를 참고해주세요.

katej927 commented on src/apis/articles.ts:
- PR 중 해당 되는 부분
  - Changes
    - api와 Interface를 한 곳(api 파일)에 몰아두기

katej927 commented on src/app/editArticle/[id]/page.tsx:
- PR 중 해당 되는 부분
  - Changes 
    - 글 생성/수정 할 때 form의 UI 컴포넌트 (articleForm) 분리 및 재사용

katej927 commented on tsconfig.json:
- PR 중 해당 되는 부분
  - Changes 
    - alias 수정

katej927 reviewed (commented):
PR에 적어둔 Changes와 Questions들이 어느 코드를 지칭하는 것인지 설명을 적어두었습니다.

Changes는 참고해주시고 Questions는 답 부탁드립니다.

katej927 commented on src/components/articleForm/index.tsx:
- PR 중 해당 되는 부분
  - Changes
    - 글 생성/수정 할 때 form의 UI 컴포넌트 (articleForm) 분리 및 재사용

f-lab-zero commented on src/app/editArticle/[id]/page.tsx:
```diff
@@ -0,0 +1,61 @@
+      return
+    }
+
+    try {
+      const res = await fetch(`${API_URL}/api/articles/${id}`, {
```
이것도 api 빼는게 좋아보여요

katej927 replied:
🙋‍♀️ 질문 : 
멘토님~ 이거 api 빼면 에러가 나요. 혹시 `${API_URL}/articles/${id}`나 `/articles/${id}`가 아닌 다른 방식이 있을까요?

f-lab-zero replied:
함수로 빼라는 의미였는데. 왜 에러가 나나요? 그냥 에러가 난다고 이야기 하시면 제대로된 답변을 드리기 힘들어요

katej927 replied:
함수로 빼라는 의미였군요!! 리팩토링때 반영할게요.

katej927 replied:
반영 완료

katej927 committed:
refactor : revised e.preventDefault() in articleForm
- it works as default so it doesn't need to put every submit

katej927 committed:
refactor : remove unused code

katej927 commented:
# Trouble Shooting

※ PR별 유의미한 트러블 슈팅을 기록하여 개인적으로 참고 하고자 적습니다. 읽지 않으셔도 괜찮습니다.


<details>
        <summary>1. server/client component와 use hook</summary>
- src/app/editArticle/[id]/page.tsx

server와 client component를 처음 사용하게 되면서 사용할 수 있는 기능에 제약이 있는 것을 잘 알지 못했다.
백엔드에서 데이터를 가져오는 로직을 위해 async/await도 써야 하고 (server component 기능) router를 사용하고 함수도 내려줘야 하는데 (client component 기능) 모두 사용하려니 계속 에러가 나서 헤맸고 매우 답답했다.

전 회사에서라면 10분내로 답 안 나오면 바로 질문하라고 해서 사수님의 도움을 받으며 해결했을테지만 멘토링에서는 멘토님이 끝까지 찾길 추천하셨고 아니면 따끔하게 지적받기에 계속 찾아봤다.
꽤 오랜 시간이 걸렸고 포기하고 멘토님께 그냥 질문드려 해결하고 싶은 순간들도 있었으나, 돌아올 답이 무엇인지 알 것도 같고 그냥 한 번 찾아보았다. 개발자가 스스로 찾는 것도 중요하기 때문이다. (회사에서는 시간 관계 상의 이유로 빠르게 질문하는 것을 권유한 것으로 안다.)
아마도 에러메세지를 쳤고 원어를 해석해내는 고통을 견디면서 stackoverflow에서 우연히 use를 사용하면 해결 가능하다는 것을 봤고 그것이 내가 알지 못한 hook이며 근래에 추가 되었다는 것을 알게 되었다.
그래서 use hook에 대해 공식문서와 블로그에서 어떤 기능을 하는 hook인지 찾아보았고 내가 원하던 기능을 하는 알맞은 function임을 알았다.

뿌듯했고 성취감을 느꼈다. 얼마 전, preview 구현하면서도 이번에도 (비록 오랜 시간이 걸렸지만) 스스로 해결해냈다.

**단순히 끝날 줄 알았던 update와 delete였는데 새로이 업데이트된 기능으로 구현하는데 차질이 많았고 스스로 해결하려다 보니 오랜 시간이 걸렸으나 뿌듯했다.**
**얻은 점은 조금씩, 스스로 문제 해결하는 방법을 찾는 것 같다는 점이다. 에러 로그를 이해하고 부족하면 에러 로그를 검색하고 (원어라 고통스럽지만) issue탭이나 stackoverflow에서 나와 비슷한 문제를 겪는 이들이 있는 것을 확인하고 그들이 어떻게 해결해내고 있는지 알 수 있다.**
**앞으로도 이런 방식을 조금 더 빨리 실행해서 스스로 해결할 때 속도를 높이는 것이 좋을 것 같다.**
</details>
